### PR TITLE
encoder: update API to support get mv

### PR DESCRIPTION
**Pull request by YuJiankang**
This CL is only used to review the get MV API
1.Currently, the picture is partitioned to 16*16 block, within the 16x16 block, each 4x4 subblock has the same MV.
2.MV value will be set to 0x8000 for correspondent intra block.
3.VideoEncMVBuffer data structure is defined to store the Motion vector.
The application should allocate a buffer and invoke getOutput(VideoEncOutputBuffer * outBuffer, VideoEncMVBuffer * MVBuffer, bool withWait) to get the coded frame and MV.

size of the MVBuffer used to store the MV should equal to 
sizeof(VAMotionVectorIntel) * 16 * picture_width_in_mbs*picture_height_in_mbs.

picture_width_in_mbs = (picture_width + 15) / 16;
picture_height_in_mbs = (picture_height + 15) / 16;

4.detailed layout of Motion vector
/**Motion vector buffer layout.
 * Motion vector output is per 4x4 block. For each 4x4 block there is a pair of past and future 
 * reference MVs as defined in VAMotionVectorIntel. Depending on Subblock partition, 
 * for the shape that is not 4x4, the MV is replicated so each 4x4 block has a pair of MVs. 
 * If only past reference is used, future MV should be ignored, and vice versa. 
 * The 16x16 block is in raster scan order, within the 16x16 block, each 4x4 block MV is ordered as below in memory. 
 *
 *                      16x16 Block        
 *        -----------------------------------------
 *        |    1    |    2    |    5    |    6    |
 *        -----------------------------------------
 *        |    3    |    4    |    7    |    8    |
 *        -----------------------------------------
 *        |    9    |    10   |    13   |    14   |
 *        -----------------------------------------
 *        |    11   |    12   |    15   |    16   |
 *        -----------------------------------------
 *
 **/

/** \brief Motion vector data structure. */
typedef struct _VAMotionVectorIntel {
    /** \mv0[0]: horizontal motion vector for past reference */
    /** \mv0[1]: vertical motion vector for past reference */
    /** \mv1[0]: horizontal motion vector for future reference */
    /** \mv1[1]: vertical motion vector for future reference */
    short  mv0[2];  /* past reference */
    short  mv1[2];  /* future reference */
} VAMotionVectorIntel;

### CODE REVIEW

File: interface/VideoEncoderDefs.h [hunk 1]
```diff
@@ -155,6 +155,18 @@ typedef struct VideoEncOutputBuffer {
 #endif
 }VideoEncOutputBuffer;
 
+#ifdef __BUILD_GET_MV__
+typedef struct VideoEncMVBuffer {
+    uint8_t *data;
```
Comment (xuguangxin) [inline interface/VideoEncoderDefs.h]: How many bytes need allocated to data?Where to get it?Do we need add something like IVideoEncoder.getMaxOutSize?

Comment (xuguangxin) [inline interface/VideoEncoderDefs.h]: better have some comment here to give more details

**Comment by YuJiankang** [comment]
size of the MVBuffer used to store the MV should equal to 
sizeof(VAMotionVectorIntel) * 16 * picture_width_in_mbs * picture_height_in_mbs.

picture_width_in_mbs = (picture_width + 15) / 16;
picture_height_in_mbs = (picture_height + 15) / 16;

Update the CL to add more comment and add a API to get MVBuffer size for convenience of users

**Comment by YuJiankang** [comment]
updated the patch for the implementation and add the test example, please help review.
1. need to add "--enable-getmv" option while compiling libyami.
2. refer to encoder.cpp for the usage of the API.

File: encoder/vaapiencoder_h264.cpp [hunk 2]
```diff
@@ -1353,11 +1362,39 @@ bool VaapiEncoderH264::ensureSlices(const PicturePtr& picture)
...
+        if (!picture->addMVBuffer(size))
+            return ret;
+
+        bufID = picture->getMVBufferID();
+        fei_pic_param.mv_data = bufID;
+        fei_pic_param.function = VA_ENC_FUNCTION_ENC_PAK_INTEL;
+        fei_pic_param.num_mv_predictors   = 1;
+
+        size = sizeof(VAEncMiscParameterBuffer) + sizeof(VAEncMiscParameterFEIFrameControlH264Intel);
+        if (!picture->addFEIBuffer(size))
```
Comment (xuguangxin) [inline encoder/vaapiencoder_h264.cpp]: can we use VaapiEncPicture::newMisc for this ?

File: encoder/vaapiencoder_h264.cpp [hunk 3]
```diff
@@ -1353,11 +1362,39 @@ bool VaapiEncoderH264::ensureSlices(const PicturePtr& picture)
...
+        VAEncMiscParameterFEIFrameControlH264Intel *misc_fei_pic_control_param;
+        VAEncMiscParameterFEIFrameControlH264Intel   fei_pic_param;
+        getMVBufferSize(&size);
+        if (!picture->addMVBuffer(size))
+            return ret;
+
+        bufID = picture->getMVBufferID();
+        fei_pic_param.mv_data = bufID;
+        fei_pic_param.function = VA_ENC_FUNCTION_ENC_PAK_INTEL;
+        fei_pic_param.num_mv_predictors   = 1;
```
Comment (xuguangxin) [inline encoder/vaapiencoder_h264.cpp]: why it's only one?

File: encoder/vaapiencoder_base.cpp [hunk 4]
```diff
@@ -417,6 +427,46 @@ Encode_Status VaapiEncoderBase::getOutput(VideoEncOutputBuffer * outBuffer, bool
     return ENCODE_SUCCESS;
 }
 
+#else
+Encode_Status VaapiEncoderBase::getOutput(VideoEncOutputBuffer * outBuffer, VideoEncMVBuffer * MVBuffer, bool withWait)
+{
+    unsigned char *data = NULL;
```
Comment (xuguangxin) [inline encoder/vaapiencoder_base.cpp]: is any possible to use getOutput(VideoEncOutputBuffer * outBuffer, bool withWait) ? or we can move common code to other function.

File: encoder/vaapiencpicture.cpp [hunk 5]
```diff
@@ -89,4 +92,32 @@ Encode_Status VaapiEncPicture::getOutput(VideoEncOutputBuffer * outBuffer)
     return ENCODE_SUCCESS;
 }
 
+#ifdef __BUILD_GET_MV__
+bool VaapiEncPicture::addMVBuffer(uint32_t size)
+{
+    m_MVBuffer = createBufferObject(VAEncFEIMVBufferTypeIntel, size, NULL, NULL);
+    return m_MVBuffer != NULL;
+}
+void *VaapiEncPicture::mapMVBuffer(uint32_t *mappedSize)
```
Comment (xuguangxin) [inline encoder/vaapiencpicture.cpp]: better use this prototype.
bool editMvBuffer(void*& buffer, uint32_t size);

File: encoder/vaapiencoder_h264.cpp [hunk 6]
```diff
@@ -1353,11 +1362,39 @@ bool VaapiEncoderH264::ensureSlices(const PicturePtr& picture)
...
 {
     Encode_Status ret = ENCODE_FAIL;
+
     SurfacePtr reconstruct = createSurface();
     if (!reconstruct)
         return ret;
     {
         AutoLock locker(m_paramLock);
+
+#ifdef __BUILD_GET_MV__
```
Comment (xuguangxin) [inline encoder/vaapiencoder_h264.cpp]: can we hiddle all this details in bool editMvBuffer(void*& buffer, uint32_t size);?

File: encoder/vaapiencpicture.cpp [hunk 7]
```diff
@@ -89,4 +92,32 @@ Encode_Status VaapiEncPicture::getOutput(VideoEncOutputBuffer * outBuffer)
     return ENCODE_SUCCESS;
 }
 
+#ifdef __BUILD_GET_MV__
+bool VaapiEncPicture::addMVBuffer(uint32_t size)
```
Comment (xuguangxin) [inline encoder/vaapiencpicture.cpp]: no need this we use 
bool editMvBuffer(void*& buffer, uint32_t size);

File: encoder/vaapiencpicture.cpp (hunk 5 above)
Comment (xuguangxin) [inline encoder/vaapiencpicture.cpp]: buffer == zero means no need map the mv buffer

File: tests/encodeinput.cpp [hunk 8]
```diff
@@ -349,3 +349,13 @@ bool createOutputBuffer(VideoEncOutputBuffer* outputBuffer, int maxOutSize)
     return true;
 }
 
+#ifdef __BUILD_GET_MV__
+bool createMVBuffer(VideoEncMVBuffer* MVBuffer, int Size)
+{
+    MVBuffer->data = static_cast<uint8_t*>(malloc(Size));
```
Comment (xuguangxin) [inline tests/encodeinput.cpp]: where is free.

**Comment by YuJiankang** [comment]
CL has updated according to the dicussion with Guangxin.

**Comment by xuguangxin** [comment]
+1

**Pull request merged**